pkg/openstack/routes.go

```diff
@@ -87,8 +176,35 @@ func (r *Routes) ListRoutes(ctx context.Context, clusterName string) ([]*cloudpr
 	}
 
 	routes := make([]*cloudprovider.Route, 0, len(router.Routes))
+
+	// detect router's private network ID for further VM ports filtering
+	r.networkIDs, err = getRouterNetworkIDs(r.network, r.os.routeOpts.RouterID)
+	if err != nil {
+		return nil, err
+	}
+
+	rules, err := openstackutil.GetSecurityGroupRules(r.network, rules.ListOpts{SecGroupID: r.nodeSecurityGroupId, Direction: string(rules.DirIngress)})
+	if err != nil {
+		return nil, err
+	}
+
 	for _, item := range router.Routes {
 		nodeName, foundNode := getNodeNameByAddr(item.NextHop, nodes)
+		if foundNode {
+			if !r.checkSecurityGroupRules(item, rules, nodeName, nodes) {
+				continue
+			}
+
+			// get the node port that the route next hop addr belong to
+			port, err := getPortByIP(r.network, item.NextHop, r.networkIDs)
+			if err != nil {
+				return nil, err
+			}
+
+			if !r.checkPortSecurityRules(port, item) {
+				continue
+			}
```

Comment:
Why do we do that on a `ListRoutes()` call?

Reply:
For reconciling. If the security group and related rules aren't set, the connectivity between different node can't be ensure, in this situation, CreateRoutes() should be triggered to set security group and related rules. In my opinion, it is useful when users need the feature and upgrade their cluster from previous edition to the lastest(Include this PR).

Reply:
Yeah, makes sense now, thanks!